src/main/scala/com/microsoft/kusto/spark/datasource/DefaultSource.scala

```diff
@@ -57,48 +71,76 @@ class DefaultSource extends CreatableRelationProvider
     }
   }
 
+
   override def createRelation(sqlContext: SQLContext, parameters: Map[String, String]): BaseRelation = {
     val requestedPartitions = parameters.get(KustoOptions.KUSTO_NUM_PARTITIONS)
-    val readMode = parameters.getOrElse(KustoOptions.KUSTO_READ_MODE, "scale")
+    val readMode = parameters.getOrElse(KustoOptions.KUSTO_READ_MODE, "scale").toLowerCase(Locale.ROOT)
     val partitioningMode = parameters.get(KustoOptions.KUSTO_READ_PARTITION_MODE)
-    val numPartitions = setNumPartitionsPerMode(sqlContext, requestedPartitions, readMode, partitioningMode)
+    val isLeanMode = readMode.equals("lean")
 
+    val numPartitions = setNumPartitionsPerMode(sqlContext, requestedPartitions, isLeanMode, partitioningMode)
     if (!KustoOptions.supportedReadModes.contains(readMode)) {
       throw new InvalidParameterException(s"Kusto read mode must be one of ${KustoOptions.supportedReadModes.mkString(", ")}")
     }
 
-    if (numPartitions != 1 && readMode.equals("lean")) {
+    if (numPartitions != 1 && isLeanMode) {
       throw new InvalidParameterException(s"Reading in lean mode cannot be done on multiple partitions. Requested number of partitions: $numPartitions")
     }
 
-    var storageSecreteIsAccountKey = true
-    var storageSecrete = parameters.get(KustoOptions.KUSTO_BLOB_STORAGE_ACCOUNT_KEY)
-    if (storageSecrete.isEmpty) {
-      storageSecrete = parameters.get(KustoOptions.KUSTO_BLOB_STORAGE_SAS_KEY)
-      if (storageSecrete.isDefined) storageSecreteIsAccountKey = false
+    var storageSecretIsAccountKey = true
+    var storageSecret = parameters.get(KustoOptions.KUSTO_BLOB_STORAGE_ACCOUNT_KEY)
+    if (storageSecret.isEmpty) {
+      storageSecret = parameters.get(KustoOptions.KUSTO_BLOB_STORAGE_SAS_URL)
+      if (storageSecret.isDefined) storageSecretIsAccountKey = false
+    }
+
+    if(authentication == null){
+      val parsedParams: (KustoAuthentication, KustoCoordinates, KeyVaultAuthentication) = KustoDataSourceUtils.parseSourceParameters(parameters)
+      authentication = parsedParams._1
+      kustoCoordinates = parsedParams._2
+      keyVaultAuthentication = parsedParams._3
+    }
+
+    val (kustoAuthentication, storageParameters): (KustoAuthentication, StorageParameters) = if (keyVaultAuthentication != null) {
+      // Get params from keyVault
+      if(isLeanMode){
```

Comment:
Why special for LeanMode?
Makes sense to either return all parameters, or define which parameters to fetch from KV, e.g. as bitmap: getAuthenticationParamsFromKeyVault(KUSTO_FETCH_BLOB_STORAGE_ACCOUNT_NAME_ | KUSTO_FETCH_BLOB_CONTAINER | ...) 
or (probably better getAuthenticationParamsFromKeyVault(fetchStorage: Boolean = false)

Reply:
talked F2F